### PR TITLE
Change Dataset interface to support sparse matrix.

### PR DESCRIPTION
**Pull request by eugene-yang**
Changed the `libact.base.dataset.Dataset` interface to support sparse matrix as X.
The interfaces for `get_entries`, `get_labeled_entries` and `get_unlabled_entries` are changed.
Since most of the usage of these methods are getting the list of tuple and zip them back to the a feature matrix and list of labels, directly change the interface to output in this format would benefit both using and storing the data. 

This would also directly support `scipy.sparse.csr_matrix` since the zipping during the initialization is removed. 
The interface of `Dataset.data[]` is still implemented via `__getitem__` magic method to support some of the use case that involve direct access to the entries.

### CODE REVIEW

File: libact/base/dataset.py [hunk 1]
```diff
@@ -47,17 +55,34 @@ def __len__(self):
...
+        Get the mask of labeled entries.
+
+        Returns
+        -------
+        mask: numpy array of bool, shape = (n_sample, )
+        """
+        return ~np.fromiter( ( e is None for e in self._y), dtype=bool )
 
-    def len_labeled(self):
+    def len_labeled(self):
```
Comment (skgg) [inline libact/base/dataset.py]: Is this a redundant whitespace?

Comment (eugene-yang) [reply libact/base/dataset.py]: Yep, sorry I missed that. 
Let me do an update.

**Comment by yangarbiter** [comment]
The changes looks good to my.
Thanks for the contribution.
Just fix the coding style and I think it will be ready to go.

**Comment by eugene-yang** [comment]
@yangarbiter Do you want me to fix them? 
Like the whitespaces around the brackets?

**Comment by yangarbiter** [comment]
Yes, please fix them
For example `lb = lbr.label( trn_ds.data[ask_id][0] )` should be `lb = lbr.label(trn_ds.data[ask_id][0])`
I didn't mark all of them.
But try to comply with google's style guide (https://google.github.io/styleguide/pyguide.html#36-whitespace)
Thanks.

File: libact/base/dataset.py [hunk 2]
```diff
@@ -142,44 +171,42 @@ def format_sklearn(self):
...
 
         Returns
         -------
-        labeled_entries : list of (feature, label) tuple
-            Labeled entries
+        X: numpy array or scipy matrix, shape = ( n_sample labeled, n_features )
+        y: list, shape = (n_samples lebaled)
         """
-        return list(filter(lambda entry: entry[1] is not None, self.data))
+        return self._X[self.get_labeled_mask()], self._y[self.get_labeled_mask()].tolist()
```
Comment (yangarbiter) [inline libact/base/dataset.py]: Do we need to make y into a list, can we just keep it numpy array?

File: libact/query_strategies/quire.py [hunk 3]
```diff
@@ -69,14 +69,16 @@ class QUIRE(QueryStrategy):
...
     def __init__(self, *args, **kwargs):
         super(QUIRE, self).__init__(*args, **kwargs)
-        self.Uindex = [
-            idx for idx, _ in self.dataset.get_unlabeled_entries()
-        ]
-        self.Lindex = [
-            idx for idx in range(len(self.dataset)) if idx not in self.Uindex
-        ]
+        self.Uindex = self.dataset.get_unlabeled_entries()[0].tolist()
+        self.Lindex = np.where( self.dataset.get_labeled_mask() )[0].tolist()
```
Comment (yangarbiter) [inline libact/query_strategies/quire.py]: misses a white space here

**Comment by yangarbiter** [comment]
Last three questions an I'll merge.
Thanks for the hard work @eugene-yang .

File: libact/base/dataset.py (hunk 2 above)
Comment (eugene-yang) [reply libact/base/dataset.py]: We actually need this. 
For some multi-label code, I think they are taking advantage of the nested structure of out output. I removed it ran against the unit test and got back with a lot of fails. So I would suggest just keep it as a list. And I don't think the performance would not be affected too much.

File: libact/query_strategies/quire.py (hunk 3 above)
Comment (eugene-yang) [reply libact/query_strategies/quire.py]: Good catch.

File: libact/base/dataset.py (hunk 2 above)
Comment (yangarbiter) [reply libact/base/dataset.py]: Ok, I think we can merge first and I'll look into this later.

Comment (eugene-yang) [reply libact/base/dataset.py]: Cool. I am starting to adapt libact to my own research experiments. So might start opening more full requests for improvements in the future :)

**Pull request merged**